agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -0,0 +1,233 @@
+package com.luxoft.blockchainlab.corda.hyperledger.indy
+
+import com.fasterxml.jackson.annotation.JsonProperty
+import com.fasterxml.jackson.databind.node.ObjectNode
+import com.luxoft.blockchainlab.hyperledger.indy.*
+import com.luxoft.blockchainlab.hyperledger.indy.utils.SerializationUtils
+import mu.KotlinLogging
+import org.java_websocket.client.WebSocketClient
+import org.java_websocket.handshake.ServerHandshake
+import java.lang.Thread.sleep
+import java.net.URI
+import java.time.Instant
+import java.util.*
+import kotlin.RuntimeException
+
+
+class PythonRefAgentConnection : AgentConnection {
+    override fun disconnect() {
+        if (getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTION_CONNECTED) {
+            webSocket.closeBlocking()
+            connectionStatus = AgentConnectionStatus.AGENT_CONNECTION_DISCONNECTED
+        }
+    }
+
+    override fun connect(url: String, login: String, password: String) {
+        disconnect()
+        webSocket = AgentWebSocketClient(URI(url))
+        webSocket.apply {
+            connectBlocking()
+            sendJson(StateRequest())
+            var stateResponse = waitForMessageOfType<State>(MESSAGE_TYPES.STATE_RESPONSE)
+            if(!checkState(stateResponse, login)) {
+                sendJson(WalletConnect(login, password, id = Instant.now().toEpochMilli().toString()))
+                stateResponse = waitForMessageOfType<State>(MESSAGE_TYPES.STATE_RESPONSE)
+            }
+            if(!checkState(stateResponse, login))
+                throw AgentConnectionException("Error connecting to $url")
+        }
+    }
+
+    private var connectionStatus : AgentConnectionStatus = AgentConnectionStatus.AGENT_CONNECTION_DISCONNECTED
+
+    override fun getConnectionStatus(): AgentConnectionStatus = connectionStatus
+
+    private var counterParty: IndyParty? = null
+
+    private lateinit var webSocket : AgentWebSocketClient
+
+    private fun checkState(stateMessage: State?, userName: String) : Boolean {
+        return if(stateMessage != null &&
+                stateMessage.content?.get("initialized") == true &&
+                stateMessage.content?.get("agent_name") == userName ) {
+            connectionStatus = AgentConnectionStatus.AGENT_CONNECTION_CONNECTED
+            true
+        } else {
+            connectionStatus = AgentConnectionStatus.AGENT_CONNECTION_DISCONNECTED
+            false
+        }
+    }
+
+    override fun acceptInvite(invite: String) {
+        if (invite != null && getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTION_CONNECTED) {
+            sendJson(ReceiveInviteMessage(invite))
+            val invite = webSocket.waitForMessageOfType<InviteReceivedMessage>(MESSAGE_TYPES.INVITE_RECEIVED)
+            sendRequest(invite.key)
+            val response = webSocket.waitForMessageOfType<RequestResponseReceivedMessage>(MESSAGE_TYPES.RESPONSE_RECEIVED)
```

Comment:
What happens with connection if response/invite objects are invalid? Do we need to close it or reset the state machine?

Reply:
No we need to retry, but it depends on business logic.

Reply:
Let's add "response handling" task in backlog, please.

Reply:
I would like also to provide simple md with driver constraints

Reply:
I believe it is better to cover all such stuff with negative testing.

Reply:
https://github.com/hyperledger-labs/cordentity/projects/1#card-18735683